### PR TITLE
refactor: extract shard resolution and tenant validation logic

### PR DESCRIPTION
**Pull request by salvatore-campagna-weaviate**
This PR extracts the shard resolution and tenant validation logic from the monolithic `Index` struct into two focused packages. The refactor maintains identical behavior while improving code organization and testability.

Two new packages handle the core logic that was previously embedded in `Index` methods:

**`resolver` package** implements shard resolution strategies using either consistent UUID hashing (single-tenant) or direct tenant mapping (multi-tenant). It includes optimized batch processing with grouping utilities and preserves the exact same shard assignment behavior as the original code.

**`multitenancy` package** provides tenant validation strategies that reject tenant parameters in single-tenant mode and validate tenant existence/activity in multi-tenant mode. Bulk validation reduces schema round trips through deduplication and batched lookups.

Tenant validation and shard resolution must happen together because shard resolution cannot proceed without valid tenant context. In the original `Index` code, methods like `putObjectBatch` perform these operations together - validation failure immediately stops processing, and both operations share the same schema data.

The design uses function pointers rather than interfaces, following idiomatic Go principles. The Go proverb **"accept interfaces, return structs"** guides this approach - callers shouldn't depend on abstractions they don't control. By returning concrete types that embed function pointers, we eliminate runtime interface dispatch overhead while keeping the **strategy selection at compile time**. This also prevents the common anti-pattern of returning interfaces from constructors, which creates unnecessary coupling and makes testing more complex.

Both packages include comprehensive test coverage with table-driven tests for specific scenarios and property-based tests for edge cases. The resolver tests verify distribution properties across random shard configurations, while multitenancy tests cover all validation failure modes.

A follow-up PR will integrate these packages into the main `Index` implementation, replacing the inline logic. The integration should be straightforward since the extracted packages maintain the same interfaces and behavior as the original code.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/sharding/resolver.go [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+// Validation errors are passed through unchanged to enable mapping to specific
+// HTTP status codes, while parsing errors are wrapped with context.
+//
+// Parameters:
+//   - ctx: request context for validation operations
+//   - object: the object to resolve a shard for
+//
+// Returns the ShardTarget containing the shard name and object, or an error
+// if validation fails or UUID parsing fails.
+func (r *byUUIDShardResolver) ResolveShard(ctx context.Context, object *storobj.Object) (ShardTarget, error) {
```
Comment (antas-marcin) [inline adapters/repos/db/sharding/resolver.go]: can we change the return type to a pointer to a struct `*ShardTarget`?

Comment (salvatore-campagna-weaviate) [reply adapters/repos/db/sharding/resolver.go]: Done
